leo_segmentation/data.py

```diff
@@ -1 +1,137 @@
+        self._data_loader = DataLoader(self.dataset, batch_size=None, shuffle=False,\
+                            num_workers=0, collate_fn=self.collation_fn)
+
+    def get_batch_data(self):
+        return next(iter(self._data_loader))
+
+    def collation_fn(self, data):
+        """
+        Collation function for data loader to generate train and val batch data
+        Args: 
+            data (tuple): contains _all_class_images dict and _image_mask_embeddings
+        
+        Returns:
+            A tuple of pytorch tensors:
+            DIMS = Image or Mask shape
+            - tr_data: (batch_size, num_classes, tr_size, DIMS): training image embeddings
+            - tr_masks: (batch_size, num_classes, tr_size, DIMS): training image masks
+            - val_data: (batch_size, num_classes, val_size, DIMS): validation image embeddings
+            - val_masks: (batch_size, num_classes, val_size, DIMS): validation image masks 
+        """
+        _all_class_images, _image_mask_embeddings = data
+        for i in range(self.config["num_tasks"]):
+            class_list = list(_all_class_images.keys())
+            tr_size = self.config["n_train_per_class"]
+            val_size = self.config["n_val_per_class"]
+            num_classes = self.config["num_classes"]
+            sample_count = (tr_size + val_size)
+            random.shuffle(class_list)
+            shuffled_list = class_list[:num_classes]
+            error_message = f"len(shuffled_list) {len(shuffled_list)} is not num_classes: {num_classes}"
+            assert len(shuffled_list) == num_classes, error_message
+            image_paths = []
+            for _, class_name in enumerate(shuffled_list):
+                all_images = _all_class_images[class_name]
+                all_images = np.random.choice(all_images, sample_count, replace=False)
+                error_message = f"{len(all_images)} == {sample_count} failed"
+                assert len(all_images) == sample_count, error_message
+                image_paths.append(all_images)
+            
+            path_array = np.array(image_paths)
+            embedding_array = np.array([[_image_mask_embeddings[image_path][0] 
+                                        for image_path in class_paths]
+                                        for class_paths in path_array])
+
+            mask_array = np.array([[_image_mask_embeddings[image_path][1] 
+                                        for image_path in class_paths]
+                                        for class_paths in path_array])
+            if i == 0:
+                batch_embeddings = np.empty((self.config["num_tasks"],) + embedding_array.shape)
+                batch_masks = np.empty((self.config["num_tasks"],) + mask_array.shape)
+            batch_embeddings[i] = embedding_array
+            batch_masks[i] = mask_array
+        
+        tr_data = batch_embeddings[:,:,:tr_size,:,:]
+        tr_data_masks = batch_masks[:,:,tr_size:,:,:]
```

Comment:
tr_data_masks = batch_masks[:, :, :tr_size, :, :] #current logic is resulting in  val_size